### PR TITLE
feat(NIST 800-53): add DMS rule to pack

### LINKED ISSUE #34: NIST.800.53 - Ensure DMS replication is not public
Check that DMS Replication is not public

*Control ID:*
AC-4

*AWS Config Rule:* 
dms-replication-not-public

*Story:*
As an AWS consultant
I would like to check if my DMS replication instance configuration is compliant (Cannot be publicly accessed)
So that I can check if my application conforms with Control ID AC-4 within NIST 800-53.

*Acceptance Criteria:*

    * Given a CDK stack with a *non-compliant DMS replication* instance
        * When NIST-503 Secure Aspects is run
        * Then the CDK stack does not deploy and the consultant receives an explanation about the non compliant DMS replication instance for Control ID AC-4 within NIST 800-53 .
    * Given a CDK stack with a *compliant DMS replication* instance and other compliant resources
        * When NIST-503 Secure Aspects is run
        * Then the CDK stack deploys and the consultant does not receive an explanation about NIST 800-53.
    * Given a CDK stack *without a DMS replication* instance and other compliant resources:
        * When NIST-503 Secure Aspects is run
        * Then the CDK stack deploys and the consultant does not receive an explanation about NIST 800-53

### PR DESCRIPTION
**Pull request by jen-rich**
resolves #34

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: RULES.md [hunk 1]
```diff
@@ -138,7 +138,8 @@ There are no warnings for this rule pack.
...
+| [NIST.800.53-DynamoDBPITREnabled](https://docs.aws.amazon.com/config/latest/developerguide/dynamodb-pitr-enabled.html)                                       | 'The DynamoDB table does not have Point-in-time Recovery enabled - (Control IDs: CP-9(b), CP-10, SI-12).                                        | The recovery maintains continuous backups of your table for the last 35 days.                                                                                                                                                                                                                                                                                 |
```
Comment (dontirun) [inline RULES.md:142]:
```suggestion
| [NIST.800.53-DynamoDBPITREnabled](https://docs.aws.amazon.com/config/latest/developerguide/dynamodb-pitr-enabled.html)                                       | The DynamoDB table does not have Point-in-time Recovery enabled - (Control IDs: CP-9(b), CP-10, SI-12).                                        | The recovery maintains continuous backups of your table for the last 35 days.                                                                                                                                                                                                                                                                                 |
```

File: RULES.md [hunk 2]
```diff
@@ -138,7 +138,8 @@ There are no warnings for this rule pack.
...
+| [NIST.800.53-DMSReplicationNotPublic](https://docs.aws.amazon.com/config/latest/developerguide/dms-replication-not-public.html)                                       | 'DMS replication is public - (Control ID: AC-4).                                        | DMS replication instances can contain sensitive information and access control is required for such accounts.                                                                                                                                                                                                                                                                                 |
```
Comment (dontirun) [inline RULES.md:141]:
```suggestion
| [NIST.800.53-DMSReplicationNotPublic](https://docs.aws.amazon.com/config/latest/developerguide/dms-replication-not-public.html)                                       | The DMS replication instance is public - (Control ID: AC-4).                                        | DMS replication instances can contain sensitive information and access control is required for such accounts.                                                                                                                                                                                                                                                                                 |
```

File: src/NIST-800-53/nist-800-53.ts [hunk 3]
```diff
@@ -45,26 +47,47 @@ export class NIST80053Checks extends NagPack {
...
-      !this.ignoreRule(ignores, 'NIST.800.53-DynamoDBPITREnabled') &&
-      !nist80053DynamoDBPITREnabled(node)
+      !this.ignoreRule(ignores, 'NIST.800.53-DMSReplicationNotPublic') &&
+      !nist80053DMSReplicationNotPublic(node)
     ) {
-      const ruleId = 'NIST.800.53-DynamoDBPITREnabled';
+      const ruleId = 'NIST.800.53-DMSReplicationNotPublic';
       const info =
-        'The DynamoDB table does not have Point-in-time Recovery enabled (Control IDs: CP-9(b), CP-10, SI-12).';
+        'The DMS replication is public (Control IDs: AC-3).';
```
Comment (dontirun) [inline src/NIST-800-53/nist-800-53.ts:61]:
```suggestion
        'The DMS replication instance is public (Control IDs: AC-3).';
```

File: src/NIST-800-53/rules/dms/index.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export { default as nist80053DMSReplicationNotPublic } from './nist800053DMSReplicationNotPublic'
```
Comment (dontirun) [inline src/NIST-800-53/rules/dms/index.ts:1]:
```suggestion
/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/
export { default as nist80053DMSReplicationNotPublic } from './nist800053DMSReplicationNotPublic' 
```

File: src/NIST-800-53/rules/dms/nist800053DMSReplicationNotPublic.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import { CfnResource, Stack } from '@aws-cdk/core';
+
+/**
+ * DMS replication instances are not public - (Control ID: AC-4)
+ * @param node the CfnResource to check
+ */
+export default function (node: CfnResource): boolean {
+  if (node instanceof CfnReplicationInstance) {
+    const publicAccess = Stack.of(node).resolve(node.publiclyAccessible);
+    if (publicAccess) {
```
Comment (dontirun) [inline src/NIST-800-53/rules/dms/nist800053DMSReplicationNotPublic.ts:15]: According to [Cfn Documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dms-replicationinstance.html#cfn-dms-replicationinstance-publiclyaccessible) if this value is `undefined` then public access in enabled by default
```suggestion
    if (!(publicAccess === false) ) {
```

File: test/NIST-800-53/NIST-800-53-DMS.test.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+*/
+
+import { SynthUtils } from '@aws-cdk/assert';
+import {
+  CfnReplicationInstance
+} from '@aws-cdk/aws-dms';
+import { Aspects, Stack } from '@aws-cdk/core';
+import { NIST80053Checks } from '../../src';
+  describe('AWS Database Migration Service (AWS DMS)', () => {
+    test('nist80053DMSReplication: - DMS replication is not public - (Control ID: AC-4)', () => {
```
Comment (dontirun) [inline test/NIST-800-53/NIST-800-53-DMS.test.ts:13]:
```suggestion
    test('nist80053DMSReplication: DMS replication instances are not public - (Control ID: AC-4)', () => {
```

File: test/NIST-800-53/NIST-800-53-DMS.test.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        })
+      );
+
+      const negative = new Stack();
+      Aspects.of(negative).add(new NIST80053Checks());
+      new CfnReplicationInstance(positive, 'rInstance', {
+        replicationInstanceClass: 'dms.t2.micro',
+        publiclyAccessible: false,
+      });
+      const messages2 = SynthUtils.synthesize(positive).messages;
```
Comment (dontirun) [inline test/NIST-800-53/NIST-800-53-DMS.test.ts:35]:
```suggestion
      const messages2 = SynthUtils.synthesize(negative).messages;
```

File: test/NIST-800-53/NIST-800-53-DMS.test.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+} from '@aws-cdk/aws-dms';
+import { Aspects, Stack } from '@aws-cdk/core';
+import { NIST80053Checks } from '../../src';
+  describe('AWS Database Migration Service (AWS DMS)', () => {
+    test('nist80053DMSReplication: - DMS replication is not public - (Control ID: AC-4)', () => {
+      const positive = new Stack();
+      Aspects.of(positive).add(new NIST80053Checks());
+      new CfnReplicationInstance(positive, 'rInstance', {
+        replicationInstanceClass: 'dms.t2.micro',
+        publiclyAccessible: true,
```
Comment (dontirun) [inline test/NIST-800-53/NIST-800-53-DMS.test.ts:18]:
```suggestion
```

File: test/NIST-800-53/NIST-800-53-DMS.test.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      );
+
+      const negative = new Stack();
+      Aspects.of(negative).add(new NIST80053Checks());
+      new CfnReplicationInstance(positive, 'rInstance', {
+        replicationInstanceClass: 'dms.t2.micro',
+        publiclyAccessible: false,
+      });
+      const messages2 = SynthUtils.synthesize(positive).messages;
+      expect(messages2).toContainEqual(
```
Comment (dontirun) [inline test/NIST-800-53/NIST-800-53-DMS.test.ts:36]:
```suggestion
      expect(messages2).not.toContainEqual(
```

**Review by dontirun — CHANGES_REQUESTED** [review]
Additionally run the `npx projen build`or `npx projen eslint` command locally so that the build does not fail on linting

File: test/NIST-800-53/NIST-800-53-DMS.test.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        expect.objectContaining({
+          entry: expect.objectContaining({
+            data: expect.stringContaining('NIST.800.53-DMSReplicationNotPublic:'),
+          }),
+        })
+      );
+
+      const negative = new Stack();
+      Aspects.of(negative).add(new NIST80053Checks());
+      new CfnReplicationInstance(positive, 'rInstance', {
```
Comment (dontirun) [inline test/NIST-800-53/NIST-800-53-DMS.test.ts:31]:
```suggestion
      new CfnReplicationInstance(negative, 'rInstance', {
```

**Comment by jen-rich** [comment]
resolves #34

File: src/NIST-800-53/nist-800-53.ts [hunk 11]
```diff
@@ -44,6 +46,26 @@ export class NIST80053Checks extends NagPack {
...
+   * @param node the IConstruct to evaluate
+   * @param ignores list of ignores for the resource
+   */
+  private checkDMS(node: CfnResource, ignores: any) {
+    if (
+      !this.ignoreRule(ignores, 'NIST.800.53-DMSReplicationNotPublic') &&
+      !nist80053DMSReplicationNotPublic(node)
+    ) {
+      const ruleId = 'NIST.800.53-DMSReplicationNotPublic';
+      const info = 'The DMS replication is public (Control IDs: AC-3).';
```
Comment (dontirun) [inline src/NIST-800-53/nist-800-53.ts:60]:
```suggestion
      const info = 'The DMS replication instance is public (Control IDs: AC-3).';
```

File: src/NIST-800-53/rules/dms/index.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export { default as nist80053DMSReplicationNotPublic } from './nist800053DMSReplicationNotPublic';
```
Comment (dontirun) [inline src/NIST-800-53/rules/dms/index.ts:1]:
```suggestion
/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/
export { default as nist80053DMSReplicationNotPublic } from './nist800053DMSReplicationNotPublic';
```

File: test/NIST-800-53/NIST-800-53-DMS.test.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+import { SynthUtils } from '@aws-cdk/assert';
+import { CfnReplicationInstance } from '@aws-cdk/aws-dms';
+import { Aspects, Stack } from '@aws-cdk/core';
+import { NIST80053Checks } from '../../src';
+describe('AWS Database Migration Service (AWS DMS)', () => {
+  test('nist80053DMSReplication: - DMS replication is not public - (Control ID: AC-4)', () => {
```
Comment (dontirun) [inline test/NIST-800-53/NIST-800-53-DMS.test.ts:11]:
```suggestion
  test('nist80053DMSReplication: DMS replication instances are not public - (Control ID: AC-4)', () => {
```

**Review by dontirun — CHANGES_REQUESTED** [review]

**Review by dontirun — CHANGES_REQUESTED** [review]
Fixes in some of the documentation changes

**Review by dontirun — CHANGES_REQUESTED** [review]
Fixes in some of the documentation changes

**Review by dontirun — APPROVED** [review]

**Pull request merged**